contracts/token-vesting.clar

```diff
@@ -5,12 +5,48 @@
 ;; Withdraws are unlocked to a specified set of addresses
 ;; at a certain schedule and conditions.
 
-;; Treasury map. Stores a "depositor" => "vesting" relation.
+;; SIP-010 token trait.
+(use-trait ft-trait .sip-010-trait.sip-010-trait)
+
+;;
+;; -- Data
+;;
+
+;; Treasury map.
+;; Identify and stores vestings for the contract.
 (define-map treasury
-  principal {token: (optional principal), amount: uint})
+  uint {sender: principal, amount: uint, token-contract: principal})
+
+;; Vesting ID.
+(define-data-var vesting-id uint u0)
+
+;; Vesting Vault.
+(define-data-var vesting-vault principal 'ST1SJ3DTE5DN7X54YDH5D64R3BCB6A2AG2ZQ8YPD5)
+
+;;
+;; -- Public
+;;
 
 ;; Deposit function.
-(define-public (deposit (token principal) (amount uint))
+(define-public (deposit (token <ft-trait>) (amount uint))
   (begin
-    (map-set treasury tx-sender {token: (some token), amount: amount})
+    (map-set treasury
+      (get-next-vesting-id)
+      {
+        sender: tx-sender,
+        amount: amount,
+        token-contract: (contract-of token)
+      })
+    (try! (contract-call? token transfer
+      amount tx-sender (var-get vesting-vault) none))
```

Comment:
As commented, I would set the recipient of the deposit as `(as-contract tx-sender)` (the smart contract itself) and create withdraw methods that check whether a vesting period has been passed or not.

Reply:
Done!